tmt/steps/prepare/feature.py

```diff
@@ -0,0 +1,306 @@
+            if os_release.get('VERSION_ID', '').startswith('9'):
+                return Distro.RHEL_9
+            if os_release.get('VERSION_ID', '').startswith('8'):
+                return Distro.RHEL_8
+            if os_release.get('VERSION_ID', '').startswith('7'):
+                return Distro.RHEL_7
+
+        if os_release.get('NAME') == 'CentOS Linux' and os_release.get(
+                'VERSION_ID', '').startswith('7'):
+            return Distro.CENTOS_7
+
+        if os_release.get('NAME') == 'CentOS Stream':
+            if os_release.get('VERSION_ID', '').startswith('8'):
+                return Distro.CENTOS_STREAM_8
+            if os_release.get('VERSION_ID', '').startswith('9'):
+                return Distro.CENTOS_STREAM_9
+
+        return None
+
+
+class ToggleableFeature(Feature):
+    def enable(self) -> None:
+        raise NotImplementedError
+
+    def disable(self) -> None:
+        raise NotImplementedError
+
+
+FEDORA_REPO = 'powertools'
+FEDORA_PACKAGES = ['epel-release', 'epel-next-release']
+
+CENTOS_7_PACKAGES = ['epel-release']
+
+CENTOS_STREAM_9_REPO = 'crb'
+
+RHEL_7_REPO = 'rhel-*-optional-rpms rhel-*-extras-rpms rhel-ha-for-rhel-*-server-rpms'
+RHEL_7_PACKAGES = ['https://dl.fedoraproject.org/pub/epel/epel-release-latest-7.noarch.rpm']
+
+RHEL_8_REPO = 'codeready-builder-for-rhel-8-$(arch)-rpms'
+RHEL_8_PACKAGES = ['https://dl.fedoraproject.org/pub/epel/epel-release-latest-8.noarch.rpm']
+
+RHEL_9_REPO = 'codeready-builder-for-rhel-9-$(arch)-rpms'
+RHEL_9_PACKAGES = ['https://dl.fedoraproject.org/pub/epel/epel-release-latest-9.noarch.rpm']
+
+
+class EPEL(ToggleableFeature):
+    KEY = 'epel'
+
+    def enable(self) -> None:
+        distro = self.guest_distro
+        sudo = self.guest_sudo
+
+        if distro == Distro.FEDORA:
+            self.info('Nothing to do on Fedora for EPEL')
+        elif distro == Distro.CENTOS_7:
+            # yum install epel-release
+            self.info('Enable EPEL on CentOS 7')
+            self.guest.execute(
+                ShellScript(f'{sudo} yum -y install {" ".join(CENTOS_7_PACKAGES)}'),
+                silent=True)
+        elif distro == Distro.CENTOS_STREAM_8:
+            # dnf config-manager --set-enabled powertools
+            # dnf -y install epel-release epel-next-release
+            self.info('Enable EPEL on CentOS Stream 8')
+            self.guest.execute(
+                ShellScript(f'{sudo} dnf config-manager --set-enabled {FEDORA_REPO}')
+                & ShellScript(f'{sudo} dnf -y install {" ".join(FEDORA_PACKAGES)}'),
+                silent=True)
+        elif distro == Distro.CENTOS_STREAM_9:
+            # dnf config-manager --set-enabled crb
+            # dnf -y install epel-release epel-next-release
+            self.info('Enable EPEL on CentOS Stream 9')
+            self.guest.execute(
+                ShellScript(f'{sudo} dnf config-manager --set-enabled {CENTOS_STREAM_9_REPO}')
```

Comment:
I do not like the fact that additional (not EPEL) repos are being enabled (and without a notice) and more importantly that they are also being disabled.
Also, if such a repo won't be configured on a test system then the whole prepare phase would fail.

Reply:
The related code for additional repos is removed. Please noted that the additional repos are added because  of https://docs.fedoraproject.org/en-US/epel/#_quickstart

Reply:
I am aware of that. What I didn't like was that there was no message to a user that EPEL repo is being enabled. And most importantly, when disabling CRB you have also disabled EPEL without any notice (what if EPEL has been enabled before independently on CRB?).

Reply:
Hi @kkaarreell, now the `EPEL` package is installed only.  Do you agree with such a code block (let's take RHEL8 as an example)?

```python
    99	class EPEL(ToggleableFeature):
   100	    KEY = 'epel'
   101	
   102	    def enable(self) -> None:
   103	        distro = self.guest_distro
   104	        sudo = self.guest_sudo
...<snip>.....
   132	        elif distro == Distro.RHEL_8:
   133	            # dnf -y install https://dl.fedoraproject.org/pub/epel/epel-release-latest-8.noarch.rpm
   134	            self.info(f"Enable {self.KEY.upper()} on RHEL 8")
   135	            self.guest.execute(
   136	                ShellScript(f'{sudo} dnf -y install {" ".join(RHEL_8_PACKAGES)}'),
   137	                silent=True)
...<snip>.....
   146	
   147	    def disable(self) -> None:
   148	        distro = self.guest_distro
   149	        sudo = self.guest_sudo
...<snip>.....
   177	        elif distro == Distro.RHEL_8:
   178	            # dnf -y remove epel-release
   179	            self.info(f"Disable {self.KEY.upper()} on RHEL 8")
   180	            self.guest.execute(
   181	                ShellScript(f'{sudo} dnf -y remove epel-release'),
   182	                silent=True)
```

>  What I didn't like was that there was no message to a user that EPEL repo is being enabled. 

I'm not sure you'd like to want to have `silent=False`, please let me know if you want to have.

Reply:
Some systems like 1minutetip do have epel-release already installed but the repo is diabled. I think this task should not rely on epel-release being absent and either explicitly (re)enable EPEL repo after installing epel-release or maybe even skip the installation step when the package is already installed.

Reply:
Note for @idorax: `ShellScript` really runs `/bin/sh` or `/bin/bash`, i.e. a multiline script, with conditions and stuff, should definitely work. Feel free to make use of it, this should work

```python
ShellScript(f"""
set -x;

rpm -q epel-release || {sudo} dnf install -y epel-release;
{sudo} dnf config-manager enable epel-release
""")
```

Reply:
Should be covered by the latest implementation.